---
title: "Genentech RNA-seq example"
author: "Benjamin Johnson"
date: "September 22, 2015"
output: html_document
---

#This document contains the necessary R commands to analyze transcript abundance data that was generated from an RNA-seq experiment to observe differenes in *Mycobacterium tuberculosis* growing on glycerol and pyruvate at neutral and acidic pH.

There are 2 replicates for each condition:
1) Glycerol pH 7.0
2) Glycerol pH 5.7
3) Pyruvate pH 7.0
4) Pyruvate pH 5.7

The files are kept in the Documents folder inside the project folder (example-RNAseq) and then within a subfolder called transcript-abund-data. Thus, in order to re-run this analysis, you will want to download these files and place them in the same directory structure (e.g. in your Documents folder inside a project folder called example-RNAseq within a subfolder called transcript-abund-data).

The data can be downloaded from the website genentech-rnaseq-example.readthedocs.org

There are 8 files:
1) mapgly7a.counts
2) mapgly7b.counts
3) mapgly57a.counts
4) mapgly57b.counts
5) mappyr7a.counts
6) mappyr7b.counts
7) mappyr57a.counts
8) mappyr57b.counts

Loading the libraries for differential expression (edgeR) and heatmap plotting (gplots and RColorBrewer) Note: you will need to download and install these packages in order for the analysis to proceed.
```{r}
library("edgeR")
library("gplots")
library("RColorBrewer")
```

Read the data in from Documents/example-RNAseq/transcript-abund-data using the function called read.table() and specifying that the argument row.names=1.
```{r}
gly7a <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mapgly7a.counts", row.names=1)
gly7b <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mapgly7b.counts", row.names=1)
gly5a <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mapgly57a.counts", row.names=1)
gly5b <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mapgly57b.counts", row.names=1)
pyr7a <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mappyr7a.counts", row.names=1)
pyr7b <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mappyr7b.counts", row.names=1)
pyr5a <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mappyr57a.counts", row.names=1)
pyr5b <- read.table("~/Documents/example-RNAseq/transcript-abund-data/mappyr57b.counts", row.names=1)
```

Rename the column headings for each sample using the function colnames(). Usage: colnames(sample_variable) <- "SampleName"
```{r}
colnames(gly7a) <- "Glycerol7a"
colnames(gly7b) <- "Glycerol7b"
colnames(gly5a) <- "Glycerol5a"
colnames(gly5b) <- "Glycerol5b"
colnames(pyr7a) <- "Pyruvate7a"
colnames(pyr7b) <- "Pyruvate7b"
colnames(pyr5a) <- "Pyruvate5a"
colnames(pyr5b) <- "Pyruvate5b"
```

Aggregate the transcript abundance data using the function cbind(), which as you may have guessed is short for column bind, based on which comparison you want to make. Since we were interested in what happens at acidic pH given a carbon source, we are using the pH 7 conditions as our primary "reference" treatment.
```{r}
gly7_vs_gly5 <- cbind(gly7a, gly7b, gly5a, gly5b) #compare glycerol pH 7.0 and pH 5.7
gly7_vs_pyr5 <- cbind(gly7a, gly7b, pyr5a, pyr5b) #compare glycerol pH 7.0 and pyruvate pH 5.7
pyr7_vs_pyr5 <- cbind(pyr7a, pyr7b, pyr5a, pyr5b) #compare pyruvate pH 7.0 and 5.7
gly7_vs_pyr7 <- cbind(gly7a, gly7b, pyr7a, pyr7b) #comparing the two conditions with which TB grows normally (shouldn't really be many differences)
```

Define the which samples are replicates of one another when comparing a reference condition to a treatment condition. You can pick what treatment is considered the reference and which is the experimental treatment to compare.
```{r}
group_ref_vs_trt <- c(1,1,2,2)
```

Place the data of interest into an edgeR specific data frame called a DGEList. We specify what variable contains the transcript abundance counts with "counts" and how they are grouped with "group"
```{r}
g7_vs_g5 <- DGEList(counts=gly7_vs_gly5, group=group_ref_vs_trt) #create DGEList data frame comparing glycerol pH 7 vs pH 5.7
g7_vs_p5 <- DGEList(counts=gly7_vs_pyr5, group=group_ref_vs_trt) #create DGEList data frame comparing glycerol pH 7 vs pyruvate pH 5.7
p7_vs_p5 <- DGEList(counts=pyr7_vs_pyr5, group=group_ref_vs_trt) #create DGEList data frame comparing pyruvate pH 7 vs pH 5.7
g7_vs_p7 <- DGEList(counts=gly7_vs_pyr7, group=group_ref_vs_trt) #create DGEList data frame comparing glycerol pH 7 vs pyruvate pH 7
```

Pre-filter the transcript abundance counts to get rid of very lowly expressed genes. This is ad hoc and can be iteratively adjusted. Here, cpm is a function that calculates the counts per million and the combination with rowSums will keep genes that have greater than 5 cpm in at least 2 of the samples. We say that a treatment is having a response on a gene if at 2 of the samples of cpm greater than 5.
```{r}
prefilter_g7_vs_g5 <- rowSums(cpm(g7_vs_g5)>5) >= 2 #create the filter for glycerol pH 7 vs pH 5.7
prefilter_g7_vs_p5 <- rowSums(cpm(g7_vs_p5)>5) >= 2 #create the filter for glycerol pH 7 vs pyruvate pH 5.7
prefilter_p7_vs_p5 <- rowSums(cpm(p7_vs_p5)>5) >= 2 #create the filter for pyruvate pH 7 vs pyruvate pH 5.7
prefilter_g7_vs_p7 <- rowSums(cpm(g7_vs_p7)>5) >= 2 #create the filter for glycerol pH 7 vs pyruvate pH 7

#Apply the filter to the data
g7_vs_g5 <- g7_vs_g5[prefilter_g7_vs_g5,] #applying our filter we created earlier (its a vector that corresponds to which genes meet our filter criteria)
g7_vs_p5 <- g7_vs_p5[prefilter_g7_vs_p5,] #applying our filter we created earlier (its a vector that corresponds to which genes meet our filter criteria)
p7_vs_p5 <- p7_vs_p5[prefilter_p7_vs_p5,] #applying our filter we created earlier (its a vector that corresponds to which genes meet our filter criteria)
g7_vs_p7 <- g7_vs_p7[prefilter_g7_vs_p7,] #applying our filter we created earlier (its a vector that corresponds to which genes meet our filter criteria)
```

Normalize the data based on library size
```{r}
#Re-compute the library sizes for the normalization step next
g7_vs_g5$samples$lib.size <- colSums(g7_vs_g5$counts) #reseting the library size (lib.size) for the samples based on the number of genes being examined post-filtering
g7_vs_p5$samples$lib.size <- colSums(g7_vs_p5$counts)
p7_vs_p5$samples$lib.size <- colSums(p7_vs_p5$counts)
g7_vs_p7$samples$lib.size <- colSums(g7_vs_p7$counts)

#Calculate the normalization factors
g7_vs_g5 <- calcNormFactors(g7_vs_g5)
g7_vs_p5 <- calcNormFactors(g7_vs_p5)
p7_vs_p5 <- calcNormFactors(p7_vs_p5)
g7_vs_p7 <- calcNormFactors(g7_vs_p7)

#Look at the normalization factors
#This is an important step as we can see if one library is significantly bigger or smaller than another
g7_vs_g5$samples #normalization factors for glycerol pH 7 vs glycerol pH 5.7
g7_vs_p5$samples #normalization factors for glycerol pH 7 vs glycerol pH 5.7
p7_vs_p5$samples #normalization factors for pyruvate pH 7 vs pyruvate pH 5.7
g7_vs_p7$samples #normalization factors for glycerol pH 7 vs pyruvate pH 7
```

Plot the multi-dimensional scaling plot (similar to a PCA plot) to see if our replicates group together, see if we have potential batch effects (additional counfounding variable like time), etc.
```{r}
plotMDS(g7_vs_g5)
plotMDS(g7_vs_p5)
plotMDS(p7_vs_p5)
plotMDS(g7_vs_p7)
```

Estimate the variance of our data. estimateCommonDisp() estimates the overall BCV of dataset, averaged over all genes. estimateTagwiseDisp() estimates gene-specific variance.
```{r}
g7_vs_g5 <- estimateCommonDisp(g7_vs_g5, verbose=TRUE)
g7_vs_g5 <- estimateTagwiseDisp(g7_vs_g5)

g7_vs_p5 <- estimateCommonDisp(g7_vs_p5, verbose=TRUE)
g7_vs_p5 <- estimateTagwiseDisp(g7_vs_p5)

p7_vs_p5 <- estimateCommonDisp(p7_vs_p5, verbose=TRUE)
p7_vs_p5 <- estimateTagwiseDisp(p7_vs_p5)

g7_vs_p7 <- estimateCommonDisp(g7_vs_p7, verbose=TRUE)
g7_vs_p7 <- estimateTagwiseDisp(g7_vs_p7)
```

Plot the biological coefficient of variation (BCV) plots. This is useful to see how the variances look at a per gene expression level. Usually, more lowly expressed genes will have greater variances than more highly expressed genes.
```{r}
plotBCV(g7_vs_g5, cex=1)
plotBCV(g7_vs_p5, cex=1)
plotBCV(p7_vs_p5, cex=1)
plotBCV(g7_vs_p7, cex=1)
```

Do the differential gene expression using the exactTest() function. You can't use this method if you have a batch effect. There are explicit examples in the user documentation of edgeR on how to deal with batch effects.
```{r}
et_gly7_vs_gly5 <- exactTest(g7_vs_g5)
et_gly7_vs_pyr5 <- exactTest(g7_vs_p5)
et_pyr7_vs_pyr5 <- exactTest(p7_vs_p5)
et_gly7_vs_pyr7 <- exactTest(g7_vs_p7)
```

Look at the top most differentially expressed genes in each comparison with the topTags() function
```{r}
top_gly7_vs_gly5 <- topTags(et_gly7_vs_gly5)
top_gly7_vs_pyr5 <- topTags(et_gly7_vs_pyr5)
top_pyr7_vs_pyr5 <- topTags(et_pyr7_vs_pyr5)
top_gly7_vs_pyr7 <- topTags(et_gly7_vs_pyr7)
```

Plot a scatter plot with genes that are significantly differentially expressed (FDR < 0.05) highlighted in red. The argument cex=0.75 is scaling the size of the data points plotted.
```{r}
summary(de_gly7_vs_gly5 <- decideTestsDGE(et_gly7_vs_gly5)) #show how many genes are significantly differentially expressed (downregulated = -1, upregulated = 1, no differential expression = 0)
detags <- rownames(g7_vs_g5)[as.logical(de_gly7_vs_gly5)] #create a vector of TRUE or FALSE using the gene names (rownames) on whether a gene is differentially expressed (TRUE) or not (FALSE)
plotSmear(et_gly7_vs_gly5, de.tags=detags, cex=0.75) #plot the scatter plot
abline(h=c(-1, 1), col="blue") #the y-axis is log base 2 for fold-change so add a couple lines delineating 2-fold changes up and down

summary(de_gly7_vs_pyr5 <- decideTestsDGE(et_gly7_vs_pyr5))
detags <- rownames(g7_vs_p5)[as.logical(de_gly7_vs_pyr5)]
plotSmear(et_gly7_vs_pyr5, de.tags=detags, cex=0.75)
abline(h=c(-1, 1), col="blue")

summary(de_pyr7_vs_pyr5 <- decideTestsDGE(et_pyr7_vs_pyr5))
detags <- rownames(p7_vs_p5)[as.logical(de_pyr7_vs_pyr5)]
plotSmear(et_pyr7_vs_pyr5, de.tags=detags, cex=0.75)
abline(h=c(-1, 1), col="blue")

summary(de_gly7_vs_pyr7 <- decideTestsDGE(et_gly7_vs_pyr7))
detags <- rownames(g7_vs_p7)[as.logical(de_gly7_vs_pyr7)]
plotSmear(et_gly7_vs_pyr7, de.tags=detags, cex=0.75)
abline(h=c(-1, 1), col="blue")
```

Export the data to a comma separated file (csv). Create a new folder in the Documents > example-RNAseq called analysis-results. This is to keep our data organized and modular for analysis steps.
```{r}
FDR_gly7_vs_gly5 <- p.adjust(et_gly7_vs_gly5$table$PValue, method="BH") #create a new vector of FDR corrected p-values
gly7_vs_gly5_outfile <- cbind(cpm(g7_vs_g5), et_gly7_vs_gly5$table, FDR_gly7_vs_gly5) #create a new dataframe with cpm (transcript abundances), summary data from the differential gene expression (exactTest()) table, and FDR corrected p-values
write.csv(gly7_vs_gly5_outfile, file="~/Documents/example-RNAseq/analysis-results/gly7_vs_gly5results.csv") #write the dataframe we just made out to a folder in Documents > example-RNAseq > analysis-results

FDR_gly7_vs_pyr5 <- p.adjust(et_gly7_vs_pyr5$table$PValue, method="BH")
gly7_vs_pyr5_outfile <- cbind(cpm(g7_vs_p5), et_gly7_vs_pyr5$table, FDR_gly7_vs_pyr5)
write.csv(gly7_vs_pyr5_outfile, file="~/Documents/example-RNAseq/analysis-results/gly7_vs_pyr5results.csv")

FDR_pyr7_vs_pyr5 <- p.adjust(et_pyr7_vs_pyr5$table$PValue, method="BH")
pyr7_vs_pyr5_outfile <- cbind(cpm(p7_vs_p5), et_pyr7_vs_pyr5$table, FDR_pyr7_vs_pyr5)
write.csv(pyr7_vs_pyr5_outfile, file="~/Documents/example-RNAseq/analysis-results/pyr7_vs_pyr5results.csv")

FDR_gly7_vs_pyr7 <- p.adjust(et_gly7_vs_pyr7$table$PValue, method="BH")
gly7_vs_pyr7_outfile <- cbind(cpm(g7_vs_p7), et_gly7_vs_pyr7$table, FDR_gly7_vs_pyr7)
write.csv(gly7_vs_pyr7_outfile, file="~/Documents/example-RNAseq/analysis-results/gly7_vs_pyr7results.csv")

```